### PR TITLE
opt: support SRFs in the FROM clause

### PR DESCRIPTION
**Pull request by rytaft**
This commit adds support for SRFs and scalar functions in the `FROM` clause by
adding a new relational operator called `Zip`.

`Zip` represents a functional zip over iterators a,b,c, which returns tuples of
values from a,b,c picked "simultaneously". NULLs are used when an iterator is
"shorter" than another. For example:

    zip([1,2,3], ['a','b']) = [(1,'a'), (2,'b'), (3, null)]

Given an SRF or scalar function in the `FROM` clause, the optbuilder
builds a relational `Zip` operator with the function(s) as iterators. The optimizer
calculates basic logical properties and statistics for the `Zip`, but currently does
not attempt to optimize the expression. The exec builder converts a `Zip` operator
back to a `projectSetNode` so it can be executed.

A subsequent PR will support SRFs in the `SELECT` list.

Release note: None

### CODE REVIEW

**Review by knz — COMMENTED** [review]
Thanks for asking me to review this. The code looks good, yet I would like to understand better: why implement a standalone relational operator `zip` when semantic analysis only produces `projectSet` at the head of the optimization pipeline?

Are you planning to semantically desugar uses of SRF to an actual correlated join with a zip on the right in the relational expression?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27306#-:-LH2whCDCDowqC45o5ES:b-ws7hqw)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Review by rytaft — COMMENTED** [review]
Hi @knz, thanks for taking a look.

> Are you planning to semantically desugar uses of SRF to an actual correlated join with a zip on the right in the relational expression?

Exactly. I think using the existing `InnerJoinApply` operator in the optimizer will give us more flexibility to optimize the expression. For example, if we find that the `Zip` is not correlated with the left hand side, we can convert the `InnerJoinApply` to a regular `InnerJoin`.

If we can't remove the `InnerJoinApply`, the execbuilder will detect that there is an `InnerJoinApply` with a `Zip`, and convert the entire combination to a `projectSetNode`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27306#-:-LH3WJO12-MIx72k_g_q:b-eqo3yl)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Review by knz — CHANGES_REQUESTED** [review]
> Reviewed 15 of 15 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27306)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt_exec_factory.go, line 519 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27306#-LH3YP5p5V1AhgwTu3bq:-LH3YP5q0QjCljVOLbe-:b19n3rs) ([raw file](https://github.com/cockroachdb/cockroach/blob/68f82479fdccc74ddf12896c88c77dee2a01a11e/pkg/sql/opt_exec_factory.go#L519)):*
> ```Go
> 			p.funcs[i] = tFunc
> 			p.numColsPerGen[i] = len(tType.Types)
> 		}
> ```

`else { p.numColsPerGen[i] = 1 }`

You should add a planning test with `... from rows from (cos(1))` or something similar to exercise this.
___
*[pkg/sql/opt/exec/execbuilder/relational_builder.go, line 691 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27306#-LH3YeDgEoBO2KbGCj1d:-LH3YeDgEoBO2KbGCj1e:b6v0bb4) ([raw file](https://github.com/cockroachdb/cockroach/blob/68f82479fdccc74ddf12896c88c77dee2a01a11e/pkg/sql/opt/exec/execbuilder/relational_builder.go#L691)):*
> ```Go
> 	}
> 
> 	node, err := b.factory.ConstructProjectSet(nil, exprs, resultCols)
> ```

Add a TODO here to explain what you said in the review comment about what remains to be done.
___
*[pkg/sql/opt/optbuilder/project_set.go, line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27306#-LH3ZGs33tNeSm4nRlz1:-LH3ZGs40oqrZ0QlPBx3:b-nq94hk) ([raw file](https://github.com/cockroachdb/cockroach/blob/68f82479fdccc74ddf12896c88c77dee2a01a11e/pkg/sql/opt/optbuilder/project_set.go#L38)):*
> ```Go
> 	// Set the flag to allow generator functions.
> 	// TODO(rytaft): remove this once we fully support SRFs.
> 	inScope.allowGeneratorFunc = true
> ```

Are you sure about this?

How does this handle e.g. `select * from rows from (generate_series(generate_series(1,2),3))`?

(pg says `ERROR:  set-returning functions must appear at top level of FROM` here)

 I think the test about where generators should be allowed must be done in some other way.

**Review by RaduBerinde — CHANGES_REQUESTED** [review]
LGTM, nice work!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27306#-:-LH3ka8v4tQf5RODRJs5:bhg6na2)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt/exec/factory.go, line 137 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27306#-LH3iyTo-yAlC_afj8w5:-LH3iyTo-yAlC_afj8w6:bfiw0i9) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b149ed1dee33290571f7f8207fc215b0c9bf8f1/pkg/sql/opt/exec/factory.go#L137)):*
> ```Go
> 	// between the output of the given node and the functional zip of the given
> 	// expressions.
> 	ConstructProjectSet(n Node, exprs tree.TypedExprs, cols sqlbase.ResultColumns) (Node, error)
> ```

Mention that the node can be `nil`, equivalent to a node that produces a single empty row.

An alternative (IMO preferable) to having this special value is to use `ConstructValues([][]tree.TypedExpr{{}}, nil)` to construct the unaryNode separately.
___
*[pkg/sql/opt/ops/relational.opt, line 376 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27306#-LH3fG4I1CXKVok83auV:-LH3fG4JBBdrPENgvV1t:b-38zd2x) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b149ed1dee33290571f7f8207fc215b0c9bf8f1/pkg/sql/opt/ops/relational.opt#L376)):*
> ```
> #    zip([1,2,3], ['a','b']) = [(1,'a'), (2,'b'), (3, null)]
> #
> # Each element of Elems is an iterator. Zip is most commonly used with
> ```

"iterator" needs a bit more explanation, it's not something that's defined in these `.opt` files. Are they always FunctionOps? I think we actually accept arbitrary scalars and treat them as a 1-result iterator?

**Review by rytaft — COMMENTED** [review]
Thanks for the reviews! Updated.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27306#-:-LH4QTFdEmQEWSmVurae:bc7d8fp)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt_exec_factory.go, line 519 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27306#-LH3YP5p5V1AhgwTu3bq:-LH3jydI31SERsdKQyxi:b-o7vuhc) ([raw file](https://github.com/cockroachdb/cockroach/blob/68f82479fdccc74ddf12896c88c77dee2a01a11e/pkg/sql/opt_exec_factory.go#L519)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`else { p.numColsPerGen[i] = 1 }`

You should add a planning test with `... from rows from (cos(1))` or something similar to exercise this.
</blockquote></details>

Good catch! Done.
___
*[pkg/sql/opt/exec/factory.go, line 137 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27306#-LH3iyTo-yAlC_afj8w5:-LH4IJ4eDqwULascRMvh:bqsc47l) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b149ed1dee33290571f7f8207fc215b0c9bf8f1/pkg/sql/opt/exec/factory.go#L137)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Mention that the node can be `nil`, equivalent to a node that produces a single empty row.

An alternative (IMO preferable) to having this special value is to use `ConstructValues([][]tree.TypedExpr{{}}, nil)` to construct the unaryNode separately.
</blockquote></details>

Good idea - I used `ConstructValues`
___
*[pkg/sql/opt/exec/execbuilder/relational_builder.go, line 691 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27306#-LH3YeDgEoBO2KbGCj1d:-LH3k3BEAaq_0oZDwicq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/68f82479fdccc74ddf12896c88c77dee2a01a11e/pkg/sql/opt/exec/execbuilder/relational_builder.go#L691)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Add a TODO here to explain what you said in the review comment about what remains to be done.
</blockquote></details>

Done.
___
*[pkg/sql/opt/ops/relational.opt, line 376 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27306#-LH3fG4I1CXKVok83auV:-LH4OsEXEhvgCG6pUqxG:b-jyfcct) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b149ed1dee33290571f7f8207fc215b0c9bf8f1/pkg/sql/opt/ops/relational.opt#L376)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

"iterator" needs a bit more explanation, it's not something that's defined in these `.opt` files. Are they always FunctionOps? I think we actually accept arbitrary scalars and treat them as a 1-result iterator?
</blockquote></details>

I rewrote this comment to use the word "generator" instead of iterator, and added an example with an SRF and a scalar function.

I don't think it accepts arbitrary scalars. The parser throws an invalid syntax error with this query: `select * from rows from (generate_series(0, 1), 1);`
___
*[pkg/sql/opt/optbuilder/project_set.go, line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27306#-LH3ZGs33tNeSm4nRlz1:-LH4E2eb3yBMQw5u-BMv:b-dyokog) ([raw file](https://github.com/cockroachdb/cockroach/blob/68f82479fdccc74ddf12896c88c77dee2a01a11e/pkg/sql/opt/optbuilder/project_set.go#L38)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Are you sure about this?

How does this handle e.g. `select * from rows from (generate_series(generate_series(1,2),3))`?

(pg says `ERROR:  set-returning functions must appear at top level of FROM` here)

 I think the test about where generators should be allowed must be done in some other way.
</blockquote></details>

That particular query returns the error `pq: unknown signature: generate_series(tuple{int AS generate_series}, int)`. Added a test with that query.

The `allowGeneratorFunc` flag doesn't mean that generator functions will be allowed where they shouldn't be, rather it's used (temporarily) to ensure that they are *not* used in the `SELECT` list. Once I submit the next PR to support SRFs in the `SELECT` list I'll remove this flag.

**Comment by knz** [comment]
What about `select * from rows from (generate_series((generatie_series(1,10)).*, 10))

Note how I'm using (...).* to work around the typing mismatch.

What I'm explaining is that it is neither correct not robust to rely on the type system to reject a structural invalidity.
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by rytaft** [comment]
`SELECT * FROM ROWS FROM (generate_series((generate_series(1,10)).*, 10))` causes the same error: `pq: unknown signature: generate_series(tuple{int AS generate_series}, int)`

But I think I know what you're getting at: `SELECT * FROM ROWS FROM (generate_series((generate_series(1,10)).generate_series, 10))` does not yet work in the optimizer (`not yet implemented: scalar expression: *tree.ColumnAccessExpr`), but in the existing CockroachDB code path it returns the error: `pq: generate_series(): programming error: generator functions cannot be evaluated as scalars`.

It seems that the current code is not catching this error at planning time, but rather at execution time.  Do you think this should be fixed?

**Comment by knz** [comment]
Woah, good find. Yes there is a bug! I'll look into it.

As to `ColumnAccessExpr` not supported. At first I was surprised because I thought I had modified the `opt` code for it in #26628. But now I see that PR was only looking at the tuple stars so the base case still remains to be done.

(I'd like to briefly point out that you'll need to support `ColumnAccessExpr` before you can do SRFs in SELECT clauses.)

**Comment by knz** [comment]
As to the general strategy about how to assert proper uses of functions, I would recommend you look at #26425 (in particular https://github.com/cockroachdb/cockroach/pull/26425/files#diff-21e8e8e2b86dec6d7bc04ec4c71173a0R610) and apply something similar. 

This is incidentally also the general direction taken in pg's own source code.

**Review by rytaft — COMMENTED** [review]
Made a change suggested by @andy-kimball to not add prune-able Zip cols since we don't yet have any rules that actually do the pruning.

@knz, to handle nested SRFs, for now I have updated the code to fall back to the heuristic planner if it encounters a nested SRF (I do this by setting `allowGeneratorFunc` to false as soon as a function is seen). I need to make the change you suggested to the entire optbuilder, but that will be a larger change. I can use #26423 to track the issue or create a new one just for the optimizer - let me know what you suggest.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27306#-:-LHEXOh4DpAM-6gKHK29:b-unausx)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Review by knz — APPROVED** [review]
:lgtm:

@RaduBerinde you'll need to change your disposition in reviewable before becca can merge this.

> Reviewed 1 of 1 files at r3, 12 of 12 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27306#-:-LHXn_lE1zlghrYgJQal:bayrnqf)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/opt/optbuilder/scope.go, line 621 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/27306#-LHXnhxc2sWZRxVccOAs:-LHXnhxc2sWZRxVccOAt:b-onomco) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bc6dac3eaa863e62fbeaf88f58c24bccbbd9f8b/pkg/sql/opt/optbuilder/scope.go#L621)):*
> ```Go
> 		}
> 
> 		// Disallow nested SRFs.
> ```

nit/tip: Add a cross-reference with a TODO to the place where this field is set (`(*builder).buildZip()`)
___
*[pkg/sql/opt/optbuilder/srfs.go, line 52 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/27306#-LHXmlv_CMDyXMOqo9nK:-LHXmlv_CMDyXMOqo9nL:b-djhg9z) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bc6dac3eaa863e62fbeaf88f58c24bccbbd9f8b/pkg/sql/opt/optbuilder/srfs.go#L52)):*
> ```Go
> 		// expression since it is set to false once an SRF is seen to disallow
> 		// nested SRFs.
> 		// TODO(rytaft): remove this once we fully support SRFs.
> ```

A thought (not blocking this PR): you won't be able to "remove" this -- you'll need to change it into something more clever.

The issue you'll have to solve is that a single SELECT target expression can contain multiple SRFs side-by-side, e.g. `select generate_series(...)+generate_series(...)`. So your current algorithm that flips `allowGeneratorFunc` back to `false` at the first SRF will not be adequate. Perhaps you want to add a note to yourself about this, either here as a TODO or in the other github issue where you're tracking the SRF work. Perhaps #26423 or #27398 or some new issue?

**Review by rytaft — COMMENTED** [review]
Thanks! @RaduBerinde, ok to merge?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27306#-:-LHYKbRSD3VX_0Sc_l_Z:b-hnkoua)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/opt/optbuilder/scope.go, line 621 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/27306#-LHXnhxc2sWZRxVccOAs:-LHYKVC82DbV7WXI5TMl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bc6dac3eaa863e62fbeaf88f58c24bccbbd9f8b/pkg/sql/opt/optbuilder/scope.go#L621)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit/tip: Add a cross-reference with a TODO to the place where this field is set (`(*builder).buildZip()`)
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/srfs.go, line 52 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/27306#-LHXmlv_CMDyXMOqo9nK:-LHYKXPu27T6kcx-n8lG:b-7yd1q4) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bc6dac3eaa863e62fbeaf88f58c24bccbbd9f8b/pkg/sql/opt/optbuilder/srfs.go#L52)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

A thought (not blocking this PR): you won't be able to "remove" this -- you'll need to change it into something more clever.

The issue you'll have to solve is that a single SELECT target expression can contain multiple SRFs side-by-side, e.g. `select generate_series(...)+generate_series(...)`. So your current algorithm that flips `allowGeneratorFunc` back to `false` at the first SRF will not be adequate. Perhaps you want to add a note to yourself about this, either here as a TODO or in the other github issue where you're tracking the SRF work. Perhaps #26423 or #27398 or some new issue?
</blockquote></details>

Added a TODO.

**Review by RaduBerinde — APPROVED** [review]

**Comment by rytaft** [comment]
Thanks!

bors r+

**Pull request merged**
